### PR TITLE
JSON serialization for ACTUS contract terms and additions for derivative contracts

### PR DESCRIPTION
**Pull request by yveshauser**
Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Review by palas — APPROVED** [review]
Looks good to me, just a couple of lints.

One meta request: in the future, I would much prefer to limit PRs to a single story. Otherwise the diff gets huge and it tricky to understand.

Another thing I often ask related to this is: when doing large refactorings (refactorings that move a lot of code around), it is much nicer to separate the moving of code from the modifying, because otherwise it is really difficult to see what was moved and what was modified

File: marlowe-actus/src/Language/Marlowe/ACTUS/Generator/GeneratorFs.hs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+            oracle_i = let ac = context <$> constraints ct in inquiryFs ev ("_" ++ show i) date "oracle" ac
+            st_i = stateAt i sd
+            rf_i = riskFactorAt i
+            stf_i = stateToContract i <$> FS.stateTransition ev rf_i sd t st_i
+            pof_i c =
+              case payoffFs ev rf_i ct st_i sd t of
+                Nothing -> c
+                Just p  -> payoff i cf date c p
+         in stf_i >>= \stf -> let f = return . oracle_i . comment ev . stf . pof_i in f cont
```
[the remark this answers is not in the document]
Comment (yveshauser) [reply marlowe-actus/src/Language/Marlowe/ACTUS/Generator/GeneratorFs.hs]: Agree, thanks

File: marlowe-actus/src/Language/Marlowe/ACTUS/Generator/GeneratorFs.hs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+                                                                             Value (..), ValueId (ValueId))
+import           Language.Marlowe.ACTUS.Domain.BusinessEvents               (EventType (..), RiskFactors,
+                                                                             RiskFactorsMarlowe, RiskFactorsPoly (..))
+import           Language.Marlowe.ACTUS.Domain.ContractState                (ContractState, ContractStateMarlowe,
+                                                                             ContractStatePoly (..))
+import           Language.Marlowe.ACTUS.Domain.ContractTerms                (Assertion (..), AssertionContext (..),
+                                                                             Assertions (..), ContractTerms,
+                                                                             ContractTermsPoly (..), PRF (..),
+                                                                             TermValidationError (..))
+import           Language.Marlowe.ACTUS.Domain.Ops                          as O (ActusNum (..), YearFractionOps (_y))
```
[the remark this answers is not in the document]
Comment (yveshauser) [reply marlowe-actus/src/Language/Marlowe/ACTUS/Generator/GeneratorFs.hs]: Ok. In the fixed schedule module we have operations from the _Prelude_ and _Actus_. We should probably split the code accordingly.

File: marlowe-actus/src/Language/Marlowe/ACTUS/Generator/GeneratorFs.hs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+-}
+
+module Language.Marlowe.ACTUS.Generator.GeneratorFs
+  ( genFsContract )
+where
+
+import           Control.Monad.Reader
+import           Data.Foldable                                              (foldrM)
+import qualified Data.List                                                  as L (zip5)
+import           Data.Maybe                                                 (maybeToList)
```
[the remark this answers is not in the document]
Comment (yveshauser) [reply marlowe-actus/src/Language/Marlowe/ACTUS/Generator/GeneratorFs.hs:20]: _maybe_ is also exported by the Prelude

**Pull request merged**